### PR TITLE
BDD

### PR DESCRIPTION
**Pull request by tomcsmith1990**
I'm making a pull request for this just so that I can put all of my comments with the code.

### CODE REVIEW

File: BankingKataSpecs/AccountSpecs/Features/CreateAccount.feature [hunk 1]
```diff
@@ -0,0 +1,19 @@
+﻿Feature: Create account
+	In order to us my account
+	As a foo user
+	I want to be able to create an account.
+
+@mytag
+Scenario: New account has zero cash.
+	Given I want to create an account
```
Comment (tomcsmith1990) [inline BankingKataSpecs/AccountSpecs/Features/CreateAccount.feature]: This `Given` seems redundant - can it be missed out? It's equivalent to a test which doesn't require any arrangement.

File: BankingKataSpecs/AccountSpecs/Features/CreateAccount.feature [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+	When I create a new account
+	Then the account should have 0 cash
+
+Scenario: New account can't take null cash.
+	Given I want to create an account
+	And I have null cash
+	Then Creating an account should throw an argument null exception
```
Comment (tomcsmith1990) [inline BankingKataSpecs/AccountSpecs/Features/CreateAccount.feature]: This test feels very dev-oriented, due to talking about `null` and exceptions. I'm not sure whether there would be a more user focused way of writing it (`Given I have an unknown amount of cash Then creating an account should give an error`?), or whether the test fits at all. Maybe BDD doesn't lend itself well to testing failure conditions?

File: BankingKataSpecs/AccountSpecs/Features/CreateAccount.feature (hunk 1 above)
Comment (SuperDrew) [inline BankingKataSpecs/AccountSpecs/Features/CreateAccount.feature]: It is redundant, you'll see in it's step that there's a TODO: Do I need to do anything here?

File: BankingKataSpecs/AccountSpecs/Features/CreateAccount.feature (hunk 2 above)
Comment (SuperDrew) [inline BankingKataSpecs/AccountSpecs/Features/CreateAccount.feature]: I agree with it feeling dev-orientated.  I think you can test failure conditions as user stories with BDD but you're right this particular example is a bad one and may be more suitable at a lower level.

**Comment by tomcsmith1990** [comment]
I think there's a more general question of who are these tests aimed at? It's quite nice being able to read them in English, but if we (development people) are going to be the only ones using them, would we be better just writing tests in code?

File: BankingKataSpecs/AccountSpecs/Features/DepositCash.feature [hunk 3]
```diff
@@ -0,0 +1,10 @@
+﻿Feature: DepositCash
+	In order to deposit cash into my account
+	As a foo user
+	I want to be able to deposit Cash
+
+@mytag
+Scenario: Depositing my cash
+	Given I create an account with 0 cash
```
Comment (tomcsmith1990) [inline BankingKataSpecs/AccountSpecs/Features/DepositCash.feature:8]: Is there any good Resharper or other (support) for tests like these? e.g. if I had misspelled one of these conditions and want to correct it, do I have to do a find/replace on the string or can I do an automated refactoring?

File: BankingKataSpecs/AccountSpecs/Features/ToStringOverride.feature [hunk 4]
```diff
@@ -0,0 +1,9 @@
+﻿Feature: ToStringOverride
+	In order to get useful information about an account
+	As a math idiot
+	I want to be able to get an informative ToStirng for an account.
+
+@mytag
```
Comment (tomcsmith1990) [inline BankingKataSpecs/AccountSpecs/Features/ToStringOverride.feature]: What do these tags do?

File: BankingKataNSpec/DebuggerShim.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        // var types = new Type[]{typeof(Some_Type_Containg_some_Specs)};
+        var finder = new SpecFinder(types, "");
+        var builder = new ContextBuilder(finder, new Tags().Parse(tagOrClassName), new DefaultConventions());
+        var runner = new ContextRunner(builder, new ConsoleFormatter(), false);
+        ContextCollection contextCollection = builder.Contexts().Build();
+        var results = runner.Run(contextCollection);
+
+        //assert that there aren't any failures
+        results.Failures().Count().should_be(0);
+    }
```
Comment (tomcsmith1990) [inline BankingKataNSpec/DebuggerShim.cs:43]: Specflow definitely seems a lot nicer.

Comment (SuperDrew) [inline BankingKataNSpec/DebuggerShim.cs:43]: Agreed

File: BankingKataSpecs/AccountSpecs/Features/DepositCash.feature (hunk 3 above)
Comment (SuperDrew) [inline BankingKataSpecs/AccountSpecs/Features/DepositCash.feature:8]: You can bring up a list of existing steps to pick from but I haven't found any refactoring support for them.

**Comment by SuperDrew** [comment]
Good question, who are they aimed at.  There are a few purposes for these tests.  
- Executable specification.  This means that anybody can read the behavioural specification and execute them.  The link between specification and execution can show how much we've implemented and what is working.
- Show who?  Product managers, sprint review attendees, UX, research, anybody who is non-technical can understand this.  It thus can become a communication and collaboration tool.  Further to this is means people other than devs or testers can write them.  Whether we use this is up for debate.
- Link our user stories directly to executable code gives us a data point for our definition of done.
- Encourage us to write tests at a behavioural level.

We could achieve this without using something like SpecFlow but it seems to make sense to use a framework that assists us write valuable tests.  So even if only devs read the tests I still think it's useful to do although not as useful as if its used as a collaboration tool.

File: BankingKataSpecs/AccountSpecs/Features/ToStringOverride.feature (hunk 4 above)
Comment (SuperDrew) [inline BankingKataSpecs/AccountSpecs/Features/ToStringOverride.feature]: They are converted to categories in NUnit so can be used for running parts of your test suite.  @ignore tag also ignores the feature or scenario.

You can also use them for SpecFlow Hooks, for example you can have a BeforeScenario tag and hook that mean for any scenario with that tag the hook is run (this is a brief summary of research online, I've never used hooks or tags).

People also use them for tagging bug reports numbers etc...  You can have multiple tags.

**Pull request closed without merging**